### PR TITLE
Transfer ownership between objectpools

### PR DESCRIPTION
**Pull request by yixinglu**
Sometimes we need to transfer some object ownerships to another pool for releasing current pool early. e.g. in optimizer of graph, we can only save the used plan nodes in `QueryContext` when finishing optimization, other nodes could be destroyed with the release of OptGroups.

### CODE REVIEW

File: src/common/base/ObjectPool.h [hunk 1]
```diff
@@ -18,51 +18,52 @@ namespace nebula {
...
-    // Holder the ownership of the any object
-    class OwnershipHolder {
-    public:
-        template <typename T>
-        explicit OwnershipHolder(T *obj)
-            : obj_(obj), deleteFn_([](void *p) { delete reinterpret_cast<T *>(p); }) {}
-
-        ~OwnershipHolder() {
-            deleteFn_(obj_);
+    DeleteFn release(void *obj) {
```
Comment (Shylock-Hg) [inline src/common/base/ObjectPool.h:47]: I think you could move the OwernshipHolder out directly.

Comment (yixinglu) [reply src/common/base/ObjectPool.h]: It's slow to find the pointer in list.

Comment (Shylock-Hg) [reply src/common/base/ObjectPool.h]: I don't say about the container(list or map). I means you could replace list by map, but still store the OwnershipHolder for the ownership management.

**Review by laura-ding — APPROVED** [review]

**Review by Aiee — APPROVED** [review]

**Review by monadbobo — APPROVED** [review]
LGTM.

**Pull request closed without merging**